virtualizarr/readers/dmrpp.py

```diff
@@ -0,0 +1,647 @@
+        dim_tags = var_tag.findall("dap:Dim", self._ns)
+        dim_shapes = self._parse_multi_dims(dim_tags, dataset_dims)
+        # convert DAP dtype to numpy dtype
+        dtype = np.dtype(
+            self._dap_np_dtype[var_tag.tag.removeprefix("{" + self._ns["dap"] + "}")]
+        )
+        # Chunks and Filters
+        filters = None
+        shape = tuple(dim_shapes.values())
+        chunks_shape = shape
+        chunks_tag = var_tag.find("dmr:chunks", self._ns)
+        if chunks_tag is not None:
+            # Chunks
+            found_chunk_dims = self._parse_chunks_dimensions(chunks_tag)
+            chunks_shape = found_chunk_dims if found_chunk_dims is not None else shape
+            chunkmanifest = self._parse_chunks(chunks_tag, chunks_shape)
+            # Filters
+            filters = self._parse_filters(chunks_tag, dtype)
+        # Attributes
+        attrs: dict[str, Any] = {}
+        for attr_tag in var_tag.iterfind("dap:Attribute", self._ns):
+            attrs.update(self._parse_attribute(attr_tag))
+        # Remove attributes only used for parsing logic
+        fill_value = attrs.pop("_FillValue", 0.0)
+        attrs.pop("fullnamepath", None)
+        attrs.pop("origname", None)
+        attrs.pop("coordinates", None)
+        # create ManifestArray and ZArray
+        zarray = ZArray(
+            chunks=chunks_shape,
+            dtype=dtype,
+            fill_value=fill_value,
+            filters=filters,
+            order="C",
+            shape=shape,
+        )
+        marr = ManifestArray(zarray=zarray, chunkmanifest=chunkmanifest)
+        encoding = {k: attrs.get(k) for k in self._encoding_keys if k in attrs}
+        return xr.Variable(
+            dims=dim_shapes.keys(), data=marr, attrs=attrs, encoding=encoding
+        )
+
+    def _parse_attribute(self, attr_tag: ET.Element) -> dict[str, Any]:
+        """
+        Parse an attribute from a DMR attr tag. Converts the attribute value to a native python type.
+
+        Parameters
+        ----------
+        attr_tag : ET.Element
+            An ElementTree Element with an <Attr> tag.
+
+        Returns
+        -------
+        dict
+        """
+        attr: dict[str, Any] = {}
+        values = []
+        if "type" in attr_tag.attrib and attr_tag.attrib["type"] == "Container":
+            return attr
+        dtype = np.dtype(self._dap_np_dtype[attr_tag.attrib["type"]])
+        # if multiple Value tags are present, store as "key": "[v1, v2, ...]"
+        for value_tag in attr_tag:
+            # cast attribute to native python type using dmr provided dtype
+            val = (
+                dtype.type(value_tag.text).item()
+                if dtype != np.object_
+                else value_tag.text
+            )
+            if val == "*":
+                val = np.nan
+            values.append(val)
+        attr[attr_tag.attrib["name"]] = values[0] if len(values) == 1 else values
+        return attr
+
+    def _parse_filters(
```

Comment:
A lot of these methods don't technically need to be methods, because they don't use `self`. It could all be functional instead. But this isn't very important, just a note.

Reply:
Yes you're right these can be standalone. Even many of the functions that use self are just using the class constants. I will update after the initial merge